h2o-algos/src/test/java/hex/tree/TreeHandlerTest.java

```diff
@@ -0,0 +1,221 @@
+                final SharedTreeNode leftChild = sharedTreeNode.getLeftChild();
+                final SharedTreeNode rightChild = sharedTreeNode.getRightChild();
+
+                if (leftChildren[i] != -1) {
+                    assertEquals(leftChildren[i], leftChild.getNodeNumber());
+                    discoverednodes.push(sharedTreeNode.getLeftChild());
+                    nonRootNodesFound++;
+                }
+                if (rightChildren[i] != -1) {
+                    assertEquals(rightChildren[i], rightChild.getNodeNumber());
+                    discoverednodes.push(sharedTreeNode.getRightChild());
+                    nonRootNodesFound++;
+                }
+            }
+
+            assertEquals(sharedTreeSubgraph.nodesArray.size(), nonRootNodesFound + 1); // +1 for the root node that is not represented explicitely
+
+
+        } finally {
+            Scope.exit();
+            if (tfr != null) tfr.remove();
+            if (model != null) model.remove();
+        }
+    }
+
+    @Test
+    public void testSharedTreeSubgraphConversion_inclusiveLevelsIris() {
+
+        Frame tfr = null;
+        GBMModel model = null;
+
+        Scope.enter();
+        try {
+            tfr = parse_test_file("./smalldata/iris/iris2.csv");
+            DKV.put(tfr);
+            GBMModel.GBMParameters parms = new GBMModel.GBMParameters();
+            parms._train = tfr._key;
+            parms._response_column = "response";
+            parms._ntrees = 1;
+            parms._seed = 0;
+
+            model = new GBM(parms).trainModel().get();
+            final SharedTreeModel.SharedTreeOutput sharedTreeOutput = model._output;
+            assertEquals(parms._ntrees, sharedTreeOutput._ntrees);
+            assertEquals(parms._ntrees, sharedTreeOutput._treeKeys.length);
+            assertEquals(parms._ntrees, sharedTreeOutput._treeKeysAux.length);
+
+            final int treeNumber = 0;
+            final int treeClass = 0;
+            final CompressedTree auxCompressedTree = sharedTreeOutput._treeKeysAux[treeNumber][treeClass].get();
+            final SharedTreeSubgraph sharedTreeSubgraph = sharedTreeOutput._treeKeys[treeNumber][treeClass].get()
+                    .toSharedTreeSubgraph(auxCompressedTree, sharedTreeOutput._names, sharedTreeOutput._domains);
+
+            assertNotNull(sharedTreeSubgraph);
+
+            final TreeHandler.TreeProperties treeProperties = TreeHandler.convertSharedTreeSubgraph(sharedTreeSubgraph);
+            assertNotNull(treeProperties);
+
+            final String[] nodeDescriptions = treeProperties._descriptions;
+            assertEquals(sharedTreeSubgraph.nodesArray.size(), nodeDescriptions.length);
+
+            for (String nodeDescription : nodeDescriptions) {
+                assertFalse(nodeDescription.isEmpty());
+            }
+
+        } finally {
+            if (tfr != null) tfr.remove();
+            if (model != null) model.remove();
+        }
+
```

Comment:
maybe you could benefit from @Before @After: would make tests easier to the eye, without try/finally noise, and would avoid forgetting Scope.exit() like here

Reply:
In many cases (but not in all cases), Scope.exit() may be called exactly once after everything else is cleaned and done. Sometimes, not all the times. Often, you even need to work with the scope using Scope.track()...

Yet this is the pattern used all over H2O's codebase and I'd not optimize it. There is nothing to optimize. I can't think about a single time it held me back. You only extract it some place else, create two more methods with additional annotations ....And this "JUNIT before-after interceptor"  may not even be usable for all the methods in the test class.